transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -204,43 +221,23 @@ public void setIoRatio(int ioRatio) {
         this.ioRatio = ioRatio;
     }
 
-    private int epollWait(boolean oldWakenUp) throws IOException {
-        int selectCnt = 0;
-        long currentTimeNanos = System.nanoTime();
-        long selectDeadLineNanos = currentTimeNanos + delayNanos(currentTimeNanos);
-        for (;;) {
-            long timeoutMillis = (selectDeadLineNanos - currentTimeNanos + 500000L) / 1000000L;
-            if (timeoutMillis <= 0) {
-                if (selectCnt == 0) {
-                    int ready = Native.epollWait(epollFd.intValue(), events, 0);
-                    if (ready > 0) {
-                        return ready;
-                    }
-                }
-                break;
-            }
-
-            // If a task was submitted when wakenUp value was 1, the task didn't get a chance to produce wakeup event.
-            // So we need to check task queue again before calling epoll_wait. If we don't, the task might be pended
-            // until epoll_wait was timed out. It might be pended until idle timeout if IdleStateHandler existed
-            // in pipeline.
-            if (hasTasks() && WAKEN_UP_UPDATER.compareAndSet(this, 0, 1)) {
-                return Native.epollWait(epollFd.intValue(), events, 0);
-            }
+    private int epollWait(boolean oldWakeup) throws IOException {
+        // If a task was submitted when wakenUp value was 1, the task didn't get a chance to produce wakeup event.
+        // So we need to check task queue again before calling epoll_wait. If we don't, the task might be pended
+        // until epoll_wait was timed out. It might be pended until idle timeout if IdleStateHandler existed
+        // in pipeline.
+        if (oldWakeup && hasTasks()) {
+            return epollWaitNow();
+        }
 
-            int selectedKeys = Native.epollWait(epollFd.intValue(), events, (int) timeoutMillis);
-            selectCnt ++;
+        long totalDelay = delayNanos(System.nanoTime());
```

Comment:
Use `TimeUnit` to avoid time math and the magic constant?

```java
long delayInNanos = delayNanos(System.nanoTime());
int delayInSeconds = (int)min(TimeUnit.NANOSECONDS.toSeconds(delayInNanos), Integer.MAX_VALUE);
return Native.epollWait(epollFd, events, timerFd, delayInSeconds, 
    (int) min(delayInNanos - TimeUnit.SECONDS.toNanos(delayInSeconds), Integer.MAX_VALUE));
```